feat: User menu trigger plugin slots.

Kelketek opened the pull request:
This Pull request adds a plugin slot to customize the contents of the user menu activation triggers in the various headers.

**Testing instructions**:

1. Clone this repo
2. For quick testing, remove your copy of .env.development and create an env.config.jsx file. Here are some sample contents that worked with me when testing the mobile menu toggle slot:
  ```jsx
import {DIRECT_PLUGIN, PLUGIN_OPERATIONS} from '@openedx/frontend-plugin-framework';

const config = {
  ACCESS_TOKEN_COOKIE_NAME: "edx-jwt-cookie-header-payload",
  ACCOUNT_PROFILE_URL: "http://localhost:1995",
  ACCOUNT_SETTINGS_URL: "http://localhost:1997",
  BASE_URL: "localhost:8080",
  CREDENTIALS_BASE_URL: "http://localhost:8150",
  CSRF_TOKEN_API_PATH: "/csrf/api/v1/token",
  ECOMMERCE_BASE_URL: "http://localhost:8130",
  LANGUAGE_PREFERENCE_COOKIE_NAME: "openedx-language-preference",
  LMS_BASE_URL: "http://localhost:8000",
  STUDIO_BASE_URL: "http://localhost:8010",
  LOGIN_URL: "http://localhost:8000/login",
  LOGOUT_URL: "http://localhost:8000/logout",
  MARKETING_SITE_BASE_URL: "http://localhost:8000",
  ORDER_HISTORY_URL: "localhost:1996/orders",
  REFRESH_ACCESS_TOKEN_ENDPOINT: "http://localhost:8000/login_refresh",
  SEGMENT_KEY: "null",
  SITE_NAME: "Open edX",
  USER_INFO_COOKIE_NAME: "edx-user-info",
  LOGO_URL: "https://edx-cdn.org/v3/default/logo.svg",
  LOGO_TRADEMARK_URL: "https://edx-cdn.org/v3/default/logo-trademark.svg",
  LOGO_WHITE_URL: "https://edx-cdn.org/v3/default/logo-white.svg",
  FAVICON_URL: "https://edx-cdn.org/v3/default/favicon.ico",
  pluginSlots: {
    'org.openedx.frontend.layout.header_mobile_user_menu_trigger.v1': {
      keepDefault: false,
      plugins: [
        {
          op: PLUGIN_OPERATIONS.Insert,
          widget: {
            id: 'custom_mobile_user_menu_toggle',
            type: DIRECT_PLUGIN,
            RenderWidget: () => (
              <span>🦊</span>
            ),
          },
        },
      ]
    },
  },
}

export default config;
```
3. If necessary, modify your tutor dev config to allow CORS from other locations using a plugin like so:
  ```python
from tutor import hooks
hooks.Filters.ENV_PATCHES.add_item(
    (
        "openedx-lms-common-settings",
        "CORS_ORIGIN_ALLOW_ALL = True",
    )
)
```
4. Follow the instructions in the READMEs provided to test each slot.


**Author notes and concerns**:

This slot makes available the contents of these menu triggers for customization, but it notably does not let you replace them in their entirety. This is because the structure of the parent components does not allow easy overrides-- in some cases requiring child components to identify as a specific React component tag. However, this should allow for nice functionality such as being able to customize default avatars or text shown in this dropdown, or even remove the username for a more minimalist look.

openedx-webhooks commented:
Thanks for the pull request, @Kelketek!


This repository is currently maintained by `@openedx/committers-frontend`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

brian-smith-tcril commented on src/plugin-slots/DesktopUserMenuToggleSlot/README.md:
```diff
@@ -0,0 +1,77 @@
+# Learning User Menu Toggle Slot
```
```suggestion
# Desktop User Menu Toggle Slot
```

brian-smith-tcril commented on src/plugin-slots/DesktopUserMenuToggleSlot/README.md:
```diff
@@ -0,0 +1,77 @@
+
+### Slot ID: `org.openedx.frontend.layout.header_desktop_user_menu_toggle.v1`
+
+### Slot ID Aliases
+* `desktop_user_menu_toggle_slot`
```
```suggestion
```

This is a new slot so it doesn't need an alias

brian-smith-tcril commented on src/plugin-slots/DesktopUserMenuToggleSlot/README.md:
```diff
@@ -0,0 +1,77 @@
+import { faHouse } from '@fortawesome/free-solid-svg-icons';
+
+const modifyUserMenuToggle = ( widget ) => {
+  widget.content.username = "My Account";
```
This example makes me wonder if we should rename the `username` prop from something tied to the default content (the username) to something more generic. I don't have any great name ideas (first things that came to mind were "label" and "text" and I don't love either), but if people agree using a more generic term makes sense I'm sure we can think of something.

brian-smith-tcril replied:
The screenshot says "My Profile" and this text says "My Account," it'd be good to make those match.

brian-smith-tcril commented on src/plugin-slots/DesktopUserMenuToggleSlot/index.jsx:
```diff
@@ -0,0 +1,22 @@
+}) => (
+  <PluginSlot
+    id="org.openedx.frontend.layout.header_desktop_user_menu_toggle.v1"
+    idAliases={['desktop_user_menu_toggle_slot']}
```
```suggestion
```

This is a new slot so it doesn't need an alias

brian-smith-tcril commented on src/plugin-slots/LearningUserMenuToggleSlot/README.md:
```diff
@@ -0,0 +1,77 @@
+
+### Slot ID: `org.openedx.frontend.layout.header_learning_user_menu_toggle.v1`
+
+### Slot ID Aliases
+* `learning_user_menu_toggle_slot`
```
```suggestion
```

This is a new slot so it doesn't need an alias

brian-smith-tcril commented on src/plugin-slots/LearningUserMenuToggleSlot/README.md:
```diff
@@ -0,0 +1,77 @@
+
+## Description
+
+This slot is used to replace/modify/hide the contents of the user menu toggle button.
```
Maybe add something about this being the *learning* menu?

brian-smith-tcril commented on src/plugin-slots/LearningUserMenuToggleSlot/README.md:
```diff
@@ -0,0 +1,77 @@
+
+### Modify Icon
+
+The following `env.config.jsx` will modify the icon for the learning user menu toggle button. **Note:** The icon is only shown on mobile screens.
```
Not a blocker, but it'd be nice to also have a "modify" example for desktop.

brian-smith-tcril commented on src/plugin-slots/LearningUserMenuToggleSlot/index.jsx:
```diff
@@ -0,0 +1,23 @@
+}) => (
+  <PluginSlot
+    id="org.openedx.frontend.layout.header_learning_user_menu_toggle.v1"
+    idAliases={['learning_user_menu_toggle_slot']}
```
```suggestion
```

This is a new slot so it doesn't need an alias

brian-smith-tcril commented on src/plugin-slots/MobileUserMenuToggleSlot/README.md:
```diff
@@ -0,0 +1,76 @@
+# Learning User Menu Toggle Slot
```
```suggestion
# Mobile User Menu Toggle Slot
```

brian-smith-tcril commented on src/plugin-slots/MobileUserMenuToggleSlot/README.md:
```diff
@@ -0,0 +1,76 @@
+
+### Slot ID: `org.openedx.frontend.layout.header_mobile_user_menu_trigger.v1`
+
+### Slot ID Aliases
+* `learning_mobile_user_menu_trigger_slot`
```
```suggestion
```

This is a new slot so it doesn't need an alias

brian-smith-tcril commented on src/plugin-slots/MobileUserMenuToggleSlot/README.md:
```diff
@@ -0,0 +1,76 @@
+
+## Description
+
+This slot is used to replace/modify/hide the contents of the user menu toggle button.
```
It'd be good to mention something about mobile here.

brian-smith-tcril commented on src/plugin-slots/MobileUserMenuToggleSlot/README.md:
```diff
@@ -0,0 +1,76 @@
+
+### Modify Avatar
+
+The following `env.config.jsx` will modify the icon for the learning user menu toggle button. **Note:** The icon is only shown on mobile screens.
```
```suggestion
The following `env.config.jsx` will modify the icon for the mobile user menu toggle button.
```

I think mentioning the icon being mobile-only might make it *less* clear that the entire slot is only rendered on mobile (also this isn't the learning menu)

brian-smith-tcril commented on src/plugin-slots/MobileUserMenuToggleSlot/README.md:
```diff
@@ -0,0 +1,76 @@
+
+### Replace Menu toggle contents with Custom Component
+
+The following `env.config.jsx` will replace the contents of the learning user menu toggle button's contents entirely (in this case with an emoji)
```
```suggestion
The following `env.config.jsx` will replace the contents of the user menu toggle button's contents entirely (in this case with an emoji)
```

brian-smith-tcril commented on src/plugin-slots/MobileUserMenuToggleSlot/index.jsx:
```diff
@@ -0,0 +1,24 @@
+}) => (
+  <PluginSlot
+    id="org.openedx.frontend.layout.header_mobile_user_menu_trigger.v1"
+    idAliases={['learning_mobile_user_menu_trigger_slot']}
```
```suggestion
```

This is a new slot so it doesn't need an alias

brian-smith-tcril reviewed (commented):
Thank you so much for this PR!

Overall I think adding these slots is a great idea and the implementation makes a lot of sense. I left a few cleanup comments in there (mostly just some copypasta from docs) and a couple of questions.

I do want to tag @arbrandes on this to make sure we have (or plan to add) equivalent functionality in the `frontend-base` header.

Kelketek commented:
Thanks, @brian-smith-tcril and @xitij2000 . I'm anticipating updating this PR this week, and will ping you once ready :)

Kelketek commented:
@brian-smith-tcril @xitij2000 I've pushed some changes here:

1. I've removed the aliases from the new slots.
2. I've run oxipng over the png files.
3. I've fixed the noted copy-pasta issues
4. I renamed the prop from `username` to `label`. While @brian-smith-tcril 's note that this prop name might not be the best one, it does seem to be better than username to me, since if you're actually modifying that property, it's probably because you want something other than a username, yet still gives a hint of how the data is expected to be used. By contrast `text` says too *little* about what we expect that data to be used for in practice. I couldn't find a term better than either of those.

brian-smith-tcril commented on src/plugin-slots/MobileUserMenuToggleSlot/README.md:
```diff
@@ -0,0 +1,73 @@
+import { PLUGIN_OPERATIONS } from '@openedx/frontend-plugin-framework';
+
+const modifyUserMenuToggle = ( widget ) => {
+  widget.content.avatar = "https://example.com/thumbs-up.png"
```
Using `example.com` here goes against the "you can paste these examples into your own `env.config.jsx` and see what the screenshots show" strategy I've been pushing for with slot documentation. Any chance you could update this to use an image from a valid URL?

Kelketek replied:
@brian-smith-tcril I've updated this with a real-world dummy URL and updated the screenshots for consistency.

brian-smith-tcril reviewed (commented):
Overall this looks great! Thank you so much for addressing the PR comments so diligently!

As for typescript, I think it likely makes sense to land this first and then transition it to typescript in a follow-up PR. If that follow-up PR is something you'd be willing to make after this lands that'd be wonderful!

I left one tiny last comment about an example URL in there, once that's addressed I'd consider this good to go!

Thanks so much for doing this!

brian-smith-tcril reviewed (approved):
LGTM!

Thanks for working through the review process with me!

Kelketek commented:
@xitij2000 Just a ping here, if you'd like to take another look. Also, @brian-smith-tcril is there anything else needed to get this across the finish line and merged in?

Kelketek commented:
Thanks, @brian-smith-tcril !

brian-smith-tcril commented:
@Kelketek I had let it sit to make sure @xitij2000 had a chance to look at it again after the changes, but I figure we'll have a chance to address anything we may have missed in follow-up PRs.
